opm/ml/ml_model.cpp

```diff
@@ -0,0 +1,337 @@
+/*
+  Copyright (c) 2016 Robert W. Rose
+  Copyright (c) 2018 Paul Maevskikh
+  Copyright (c) 2024 NORCE
+  This file is part of the Open Porous Media project (OPM).
+
+  OPM is free software: you can redistribute it and/or modify
+  it under the terms of the GNU General Public License as published by
+  the Free Software Foundation, either version 3 of the License, or
+  (at your option) any later version.
+
+  OPM is distributed in the hope that it will be useful,
+  but WITHOUT ANY WARRANTY; without even the implied warranty of
+  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+  GNU General Public License for more details.
+
+  You should have received a copy of the GNU General Public License
+  along with OPM.  If not, see <http://www.gnu.org/licenses/>.
+*/
+
+#include <fmt/format.h>
+
+#include <cmath>
+#include <cstdio>
+#include <fstream>
+#include <limits>
+#include <opm/common/ErrorMacros.hpp>
+#include <opm/ml/ml_model.hpp>
+#include <utility>
+
+namespace Opm {
+
+template <typename T>
+bool readFile(std::ifstream& file, T& data, size_t n = 1) {
+  file.read(reinterpret_cast<char*>(&data), sizeof(T) * n);
+  return !file.fail();
+}
+
+template <class Evaluation>
+bool NNLayerActivation<Evaluation>::loadLayer(std::ifstream& file) {
+  unsigned int activation = 0;
+  OPM_ERROR_IF(!readFile<unsigned int>(file, activation),
+               "Failed to read activation type");
+
+  switch (activation) {
+  case kLinear:
+    activation_type_ = kLinear;
+    break;
+  case kRelu:
+    activation_type_ = kRelu;
+    break;
+  case kSoftPlus:
+    activation_type_ = kSoftPlus;
+    break;
+  case kHardSigmoid:
+    activation_type_ = kHardSigmoid;
+    break;
+  case kSigmoid:
+    activation_type_ = kSigmoid;
+    break;
+  case kTanh:
+    activation_type_ = kTanh;
+    break;
+  default:
+    OPM_ERROR_IF(true, fmt::format("\n Unsupported activation type "
+                                   "{}",
+                                   activation));
+  }
+
+  return true;
+}
+
+template <class Evaluation>
+bool NNLayerActivation<Evaluation>::apply(const Tensor<Evaluation>& in,
+                                          Tensor<Evaluation>& out) {
+  constexpr double sigmoid_scale = 0.2;
```

Comment:
Do not define this up here, but just before it is used (line 96). As already commented...